File: Algo/Leetcode/077Combinations.cpp
//1. DFS. O(N!)
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<int> path;
        vector<vector<int>> res;

        dfs(n,k,1,0,path,res);
        return res;
    }

    void dfs(int n, int k, int start, int cur, vector<int>& path, vector<vector<int>>& res){
        if(cur == k) res.push_back(path);
        for(int i=start;i<=n;i++){
            path.push_back(i);
            dfs(n,k,i+1,cur+1,path,res);
            path.pop_back();
        }
    }
};

// 2. bit cal
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        for(int i=0; i<pow(2,n); i++){
            if(calSum(i) == k){
                vector<int> r;
                for(int j=0; j<20; j++){
                    if(i&(1<<j)) r.push_back(j+1);
                }
                res.push_back(r);
            }
        }
        return res;
    }

    int calSum(int n){
        int count = 0;
        for(int j=0; j<32; j++)
            if(n&(1<<j)) count++;
        return count;
    }
};

File: Algo/Leetcode/077Combinations.py
class Solution(object):
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        nums = list(range(1, n+1))
        n = len(nums)
        res = []
        # for each valid combination
        for i in range(2**n):
            _res = []
            count = 0
            # check count == k
            for j in range(32):
                count += 1 if i & 1<<j else 0
            # add the nums pos
            if count == k:
                for j in range(32):
                    if i & 1<<j:
                        _res.append(nums[j])

                res.append(_res)

        return res

# dfs
class Solution(object):

    def __init__(self):
        self.res = []
        self.path = []

    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        nums = list(range(1, n + 1))
        n = len(nums)
        self.dfs(nums, 0, n, k)

        return self.res

    def dfs(self, nums, i, n, k):
        if len(self.path) == k:
            self.res.append(copy.deepcopy(self.path))
            return

        if i >= n:
            return

        for j in range(i, n):
            self.path.append(nums[j])
            self.dfs(nums, j + 1, n, k)
            self.path.pop()